language-plutus-core/test/Evaluation/DynamicBuiltins/Definition.hs

```diff
@@ -218,6 +221,81 @@ test_dynamicIdList =
                 $ mkIterApp () Plc.enumFromTo [one, ten]
         typecheckEvaluateCek env term @?= Right (EvaluationSuccess res)
 
+data TyForallStarRep text unique a
+instance (KnownTypeAst uni (TyVarRep text unique), KnownTypeAst uni a) =>
+            KnownTypeAst uni (TyForallStarRep text unique a) where
+    toTypeAst _ = case toTypeAst @uni $ Proxy @(TyVarRep text unique) of
+        TyVar () name -> TyForall () name (Type ()) . toTypeAst $ Proxy @a
+        _             -> Prelude.error "Impossible"
+
+dynamicIdRank2Name :: DynamicBuiltinName
+dynamicIdRank2Name = DynamicBuiltinName "idNil"
+
+{- Note [Higher-rank built-in functions]
+We can't unlift a monomorphic function passed to a built-in function, let alone unlift a polymorphic
+one, however we can define a built-in function that accepts an 'Opaque' term of a polymorphic type.
+However, as is always the case with an 'Opaque' term, we can't inspect it or use it in any
+non-opaque way, so a function of type
+
+    all (f :: * -> *). (all (a :: *). f a) -> all (a :: *). f a
+
+can be assigned the following meaning on the Haskell side:
+
+    \x -> x
+
+but we have no way of providing a meaning for a built-in function with the following signature:
+
+    all (f :: * -> *). all (a :: *). (all (a :: *). f a) -> f a
+
+That's because the meaning function would have to instantiate the @all (a :: *). f a@ argument
+somehow to get an @f a@, but that is exactly "using a term in a non-opaque way".
+
+Basically, since we are either generic over @term@ or, like in the example below, use 'CekValue',
+there's is no sensible way of instantiating a passed polymorphic argument (or applying a passed
+argument when it's a function, for another example).
+
+Note also that the fact that we have an @all@ after a term argument in the original type signature
```

Comment:
This section is true, but not directly relevant to a future reader of the note, I suspect.

Reply:
Ok, removed this part:

> Note also that the fact that we have an @all@ after a term argument in the original type signature
is unrelated to the discussion on whether we should allow interleaved type and term arguments as
the type scheme of that signature is "take a type, take a term, return a term" and the fact that
the final term is of a polymorphic type is irrelevant.

> However it feels more consistent to support interleaved type and term arguments if we can fake them
this way anyway (albeit in a very restrictive manner).